client/src/test/java/org/eclipse/hono/client/impl/AbstractSenderTest.java

```diff
@@ -124,6 +130,34 @@ public void testSendMessageFailsOnLackOfCredit() {
         verify(protonSender, never()).send(any(Message.class));
     }
 
+    /**
+     * Verifies that a timeout exception is triggered if no delivery update
+     * is received after the message was sent.
+     */
+    @Test
+    public void testSendMessageFailsOnTimeout() {
+
+        // GIVEN a sender that won't receive a delivery update on sending a message 
+        // and directly triggers the timeout handler
+        when(protonSender.send(any(Message.class), any(Handler.class))).thenReturn(mock(ProtonDelivery.class));
+        when(vertx.setTimer(anyLong(), any(Handler.class))).thenAnswer(invocation -> {
+            final Handler<Long> handler = invocation.getArgument(1);
+            final long timerId = 1;
+            handler.handle(timerId);
+            return timerId;
+        });
+        final AbstractSender sender = newSender("tenant", "endpoint");
+
+        // WHEN sending a message
+        final Message message = mock(Message.class);
+        final Future<ProtonDelivery> result = sender.sendMessageAndWaitForOutcome(message, NoopSpan.INSTANCE);
```

Comment:
one important thing to fix was to make sure that the span is always finished, wasn't it? If so, I think you should pass in a mocked Span and later verify that it has been finished. The same test should also be added to TelemetrySenderImplTest in order to verify if this is also the case when sending a message with QoS 0.

Reply:
Thanks for pointing that out. Implementing these tests has revealed the span was being finished twice in `AbstractSender`. I've fixes that and added the tests.